Source/DafnyCore/Resolver/ModuleResolver.cs

```diff
@@ -1262,6 +1262,15 @@ public void ResolveTopLevelDecls_Core(List<TopLevelDecl> declarations,
         FillInPostConditionsAndBodiesOfPrefixLemmas(declarations);
       }
 
+      // A function is not allowed to be used naked in its own SCC. Also, a function is not allowed to be used
+      // in any way inside a "decreases" clause its its own SCC.
+      foreach (var function in ModuleDefinition.AllFunctions(declarations)) {
+        var visitor = new DetectUnsoundFunctionReferences_Visitor(this, function);
+        visitor.Visit(function);
```

Comment:
If this visit/doDecreasesChecks/visit pattern is the only correct way to use `DetectUnsoundFunctionReferences_Visitor`, then it would be better to move this into a static method in that class, and make the constructor and DoDecreasesChecks field private.

Reply:
Good idea.